godot-core/src/builtin/collections/packed_array.rs

```diff
@@ -1064,10 +1139,83 @@ impl PackedByteArray {
     }
 }
 
+mod buffer {
+    use std::mem::MaybeUninit;
+    use std::ptr;
+
+    /// A fixed-size buffer that does not do any allocations, and can hold up to `N` elements of type `T`.
+    pub struct Buffer<T, const N: usize> {
+        buf: [MaybeUninit<T>; N],
+        len: usize,
+    }
+
+    impl<T, const N: usize> Default for Buffer<T, N> {
+        fn default() -> Self {
+            Self {
+                buf: [const { MaybeUninit::uninit() }; N],
+                len: 0,
+            }
+        }
+    }
+
+    impl<T, const N: usize> Buffer<T, N> {
+        /// Appends the given value to the buffer.
+        ///
+        /// # Panics
+        /// If the buffer is full.
+        pub fn push(&mut self, value: T) {
+            self.buf[self.len].write(value);
+            self.len += 1;
+        }
+
+        pub fn is_full(&self) -> bool {
+            self.len == N
+        }
+
+        /// Returns a slice (as a pointer and length pair) of all initialized elements in the buffer,
+        /// and sets the length of the buffer to 0.
+        ///
+        /// It is the caller's responsibility to ensure that all elements in the slice get dropped.
+        /// This must be done before adding any new elements to the buffer.
+        pub fn drain_as_slice(&mut self) -> (*const T, usize) {
+            let len = self.len;
+            self.len = 0;
+            (self.buf[0].as_ptr(), len)
+        }
+    }
+
+    impl<T, const N: usize> Drop for Buffer<T, N> {
+        fn drop(&mut self) {
+            assert!(self.len <= N);
+            if N > 0 {
+                // SAFETY: for slice_from_raw_parts_mut: self.buf[0] is a valid pointer, exactly self.len elements are initialized,
+                // and the pointer is not aliased since we have an exclusive &mut self.
+                // SAFETY: for drop_in_place: the value is valid because the slice_from_raw_parts_mut requirements are met,
+                // and there is no other way to access the value.
+                unsafe {
+                    ptr::drop_in_place(ptr::slice_from_raw_parts_mut(
+                        self.buf[0].as_mut_ptr(),
+                        self.len,
+                    ));
+                }
```

Comment:
Can you split this into 2 statements, with 2 `unsafe` blocks? Then you don't need the "for ...:" clarification.